### PR TITLE
Feat command mode

### PR DESCRIPTION
**Pull request by NickHackman**
### Features
 * infobar displays command mode input, and errors
#### Command Mode Features
     * h/help
     * q/quit
     * sort
     * ng/newglob, handles `~` and `$HOME` properly and directories are assumed to be `/dir/*`
     * df/destfolder
     * r/reverse
     * m/max

### Known bugs: 

* In command mode: cannot type `:` due to the event buffer not being cleared this character is ignored, because it will commonly be added on first loop of `get_command` in `/src/program/command_mode.rs`

* When program starts in a directory that has no images, Command mode isn't properly displayed on the infobar

### Future Fixes
 * Decouple the rendering so that I and @gurgalex can update the infobar without having to pass the info to display to `render_screen`, allow updating of smaller components like the infobar before calling render or some other solution.

* Force rendering, workaround solution being used for cases like `newglob` which should re-render when the current viewed image isn't in the new glob.

@gurgalex Here's how I'm handling it, few bugs I wanted to iron out before setting up the pull request

### CODE REVIEW

File: src/program/command_mode.rs
Comment (gurgalex) [inline src/program/command_mode.rs]: This would fail if the glob had a file or directory with a ` ` in it or escaped space even `\ `.

Example
`newglob /home/alex/My\ Pictures/Tour\ S*.png`

Comment (NickHackman) [reply src/program/command_mode.rs]: Good catch! I completely forgot about that edge case, thanks! :+1:

What do you think about the implementation of command mode as a whole? Is there a better approach that I'm missing?

Comment (gurgalex) [reply src/program/command_mode.rs]: I like this implementation better. (I mostly just want something that seems to work).

I do like that text entry doesn't depend on keycodes. (Hello, international symbols 「東方」).

Does this string expand out `ng ~/*.png`?

Shellexpand may reduce some of the hand-written code for ~ expansion.
http://docs.rs/shellexpand/1.0.0/shellexpand/

Comment (gurgalex) [reply src/program/command_mode.rs]: I really was thinking we'd need to do something for custom commands using

https://pest.rs/

or nom
https://github.com/Geal/nom/


Maybe eventually, though it'd be great? if we didn't have to.

Comment (NickHackman) [reply src/program/command_mode.rs]: > Does this string expand out `ng ~/*.png`?
> 
> Shellexpand may reduce some of the hand-written code for ~ expansion.
> http://docs.rs/shellexpand/1.0.0/shellexpand/

As of right now that doesn't work, but I think that's more of an issue with `glob` than us. I'll look into `shellexpand` thanks :smiley:

Comment (gurgalex) [reply src/program/command_mode.rs]: nice live grammar evaluator they have
Nice tools and documentation if needed though.

https://pest.rs/#editor

And a book
https://pest.rs/book/

nom still an option
https://crates.io/crates/nom

Comment (NickHackman) [reply src/program/command_mode.rs]: @Davejkane, @gurgalex potential futures issues based off of work done in command mode and discussed in #57

### Future Feature Issues
 * Infobar is easier to update/decoupled from the rest of rendering or refractored in a way where smaller components of rendering can be updated or rendered without rerendering everything
 * **Good First Issue** Info bar accepts the input of error messages, changes the current image index field to a red color and displays "Error" and then the error message in the current image path field. Update errors to now, if possible, be sent to the infobar removing "Error:" when necessary from the message
* **Good First Issue** Add the `add` command to command mode, which should follow the format `:add glob file directory` or `:a glob file directory` and append images to `self.paths.images`
* **Good First Issue** Add the `delete` command to command mode, which should follow the format `:delete /path/to/image regex` or `:d /path/to/image` otherwise delete current image from `self.paths.images`. *Don't actually delete the file*

Any other ideas? I feel like these are good first issues that require some work, but not like a crazy amount where they need to understand how the app functions as a whole.

**Comment by NickHackman** [comment]
@Davejkane, @gurgalex potential future issues based off of work done in command mode and discussed in #57

### Future Feature Issues
 * Infobar is easier to update/decoupled from the rest of rendering or refractored in a way where smaller components of rendering can be updated or rendered without rerendering everything
 * **Good First Issue** Info bar accepts the input of error messages, changes the current image index field to a red color and displays "Error" and then the error message in the current image path field. Update errors to now, if possible, be sent to the infobar removing "Error:" when necessary from the message
* **Good First Issue** Add the `add` command to command mode, which should follow the format `:add glob file directory` or `:a glob file directory` and append images to `self.paths.images`
* Add autocomplete for command mode for files, commands, etc
* Force rendering of application as mentioned above
* **Good First Issue** Keep a history of command mode commands entered by user

Any other ideas? I feel like these are good first issues that require some work, but not like a crazy amount where they need to understand how the app functions as a whole.

edit: moved from a sub comment into main discussion.

**Comment by gurgalex** [comment]
> Good First Issue Add the delete command to command mode, which should follow the format :delete /path/to/image regex or :d /path/to/image otherwise delete current image from self.paths.images. Don't actually delete the file

Maybe `untrack`?. I'd like to reserve the word `delete` or `remove` for actually deleting an mage from storage. 
Yes, the fn name currently is called `remove_image`, so possibly need to rename that sometime.

File: src/program/command_mode.rs
Comment (NickHackman) [reply src/program/command_mode.rs]: So I've looked into http://docs.rs/shellexpand/1.0.0/shellexpand/ and it uses [std::env::home_dir](https://doc.rust-lang.org/std/env/fn.home_dir.html) which is deprecated, Opened an [issue](https://github.com/netvl/shellexpand/issues/3). Afterwards it may be useful though!

**Comment by NickHackman** [comment]
Update on progress:
* Added dependency [shellexpand](https://github.com/netvl/shellexpand) to better expand `~` and `$HOME` and other env variables. Note `shellexpand` uses deprecated `std::env::home_dir` pull request sent in fixing that.
* Fixed bug, where command bar wouldn't be rendered on empty images
* Fixed bug, mentioned by @gurgalex where newglob wouldn't accept paths like `/home/etc/path\ to\ dir/*` by adding dependency [regex](https://github.com/rust-lang/regex) (will be used later on for search as well)
* Overall, restructure of code base in `src/program/command_mode.rs` to be more intuitive

Opening code for review

File: src/program/command_mode.rs
Comment (Davejkane) [inline src/program/command_mode.rs]: Please add a comment to the line. I think it's fair to say it's not self-explanatory.

File: src/program/command_mode.rs
Comment (Davejkane) [inline src/program/command_mode.rs:154]: I'm not entirely clear on what the above function does. Is it for parsing user input, specifically for the path + glob, or something else? Why is it limited to ascii? You should probably mention that in the function name, that it is specific to paths. Also, guess this won't work on windows?

File: src/infobar.rs
Comment (Davejkane) [inline src/infobar.rs]: Am I right in thinking that if I send Some() string to this function, then I'll end up in Command mode? Will this effect using the infobar for error messages?

File: src/program/command_mode.rs
Comment (Davejkane) [inline src/program/command_mode.rs]: I'm not a big fan of TODOs left in the code. Can we please deal with the things that can be dealt with and make issues for the things that need to be follow-ups.

File: src/ui.rs
Comment (Davejkane) [inline src/ui.rs]: Really not sure this holds true on non-american keyboards. Shift semicolon does not always equal colon. Please see if there's another way to do this. Maybe a virtual keycode if you can get them working.

File: src/ui.rs
Comment (Davejkane) [inline src/ui.rs]: Why are we no longer calling `process_action` on these Actions?

**Review by Davejkane — CHANGES_REQUESTED** [review]
I've made a couple of comments, nothing too major. In general I'm impressed with the code and certainly with the enthusiasm to just jump in and make such a huge feature. The project is going to be all the better for it.

File: src/ui.rs
Comment (NickHackman) [reply src/ui.rs]: Fixed in "Rewrite of Modal system, add Error"

Thanks for catching that, that was completely unintended. This was a mistake in my merging, my bad.

File: src/program/command_mode.rs
Comment (NickHackman) [reply src/program/command_mode.rs]: Fixed in "Rewrite of Modal system"

File: src/program/command_mode.rs
Comment (gurgalex) [inline src/program/command_mode.rs]: I believe this regex should be compiled only once, and not on every function call.

https://docs.rs/regex/1.1.6/regex/#example-avoid-compiling-the-same-regex-in-a-loop
Plus we could eventually have a descriptive name for every regex made.

**Review by gurgalex — CHANGES_REQUESTED** [review]
Avoid recompiling the regex every time it is called. See https://docs.rs/regex/1.1.6/regex/#example-avoid-compiling-the-same-regex-in-a-loop

File: src/infobar.rs
Comment (NickHackman) [reply src/infobar.rs]: Yeah, this was the main issue addressed in "Rewrite of Modal system".  New changes
```rust
pub enum Mode {
    /// Default mode, allows the removal, traversal, move, and copy of images
    Normal,
    /// Mode that is built off of user input, allows switching the current glob
    /// string is the input to display on the infobar
    Command(String),
    /// Mode that is meant to display errors to the user through the infobar
    /// string is the input to display on the infobar
    Error(String),
    /// Terminate condition, if this mode is set the program will stop execution
    Exit,
}
```
The `render_infobar` funciton now passes in the current mode, which contains the message if applicable and This required me to change a lot of the way I handled error handling, so now Error was added as an easy solution.

File: src/program/command_mode.rs
Comment (NickHackman) [reply src/program/command_mode.rs:154]: Added better documentation, but you're completely right my code base isn't cross platform.

This functionality will come in the next push or so :heart:

File: src/program/command_mode.rs
Comment (NickHackman) [reply src/program/command_mode.rs]: :+1:

**Comment by NickHackman** [comment]
@Davejkane thanks for the advice/review! in the future I'm going to have a more cross platform mindset compared to currently as that seems like a big issue. In addition to that a more proactive mindset as well some of the code I wrote in this was quick and dirty/sketchy. I'm going to iron out these issues ASAP :smiley: 

Yeah... this seemed like it wouldn't be too bad going in, but quickly became this behemoth.

I think that we should remove all the normal mode features from `src/program/mod.rs` -> `src/program/normal_mode.rs` or something along those lines to better layout the project, just a thought though

File: src/program/command_mode.rs
Comment (NickHackman) [reply src/program/command_mode.rs]: Yep, with the cross platform approach I'm thinking about this would be a global constant in `command_mode.rs`. 

Coming up with really good names is always hard...

Comment (gurgalex) [reply src/program/command_mode.rs]: So, how would I type this as newglob?
`/home/alex/test dir/**/*.png`

would it be
`ng /home/alex/test\ dir/**/*.png`?

Comment (NickHackman) [reply src/program/command_mode.rs]: Yeah, because otherwise it's impossible to know the end of a file name without making syscalls for every string provided and their combinations.

Funny enough, this is a bug in Neovim, and I'm not making an issue because it's impossible to fix...

make a file `~/home/$USER/test\\`
open neovim
:e test\\\ Documents/*
current buffer: `test\\ Documents/*`

This results in whenever you try to save the current buffer it fails the write operation because that's never allow to be a file which is just awesome. 

In our case we don't have that bug because of cli.rs/push_image_path just gives back nothing...

Comment (gurgalex) [reply src/program/command_mode.rs]: gotten_glob

Comment (gurgalex) [reply src/program/command_mode.rs]: Does `ng` with a directory with spaces work for you?

Comment (NickHackman) [reply src/program/command_mode.rs]: It does not, actually need to remove the `\`s in order for it to be recognized as a dir. Thanks for the catch, I completely neglected that.

File: src/ui.rs
Comment (Davejkane) [inline src/ui.rs]: Please process the action for all the actions. Otherwise if some first deletes an image, then goes to the first image, then uses the dot operator, they will delete another image, and they're probably not expecting that.

File: src/program/command_mode.rs
Comment (Davejkane) [reply src/program/command_mode.rs]: path_regex would be a fine name, and of course, it should only be compiled once, even though the overhead is presumably tiny.

**Review by Davejkane — CHANGES_REQUESTED** [review]
Still a few little things. I would like us to try to support windows if at all possible. At some point it would be nice to get a windows user on board. I have no idea if this even works on windows.

File: src/program/command_mode.rs
Comment (NickHackman) [reply src/program/command_mode.rs]: Fixed, via a large amount of rewriting 

A regex was an over complicated approach that was naive and shortsighted, for instance our input to command mode may not always a be a file. Each command should handle and parse its own input since it can be interpreted differently in this context. 

Marking as resolved.

File: src/program/command_mode.rs
Comment (NickHackman) [reply src/program/command_mode.rs:154]: Fixed in latest push, completely OS agnostic

marking as resolved

**Review by gurgalex — APPROVED** [review]

**Comment by Davejkane** [comment]
As far as I'm concerned, there's just a couple of things left. And that's the "colon is not necessarily shift+semi-colon" issue. And then there were some actions that appeared to be unprocessed and so I'm not sure if they would apply to the dot operator, but it could be that I've misunderstood a change there.

**Comment by gurgalex** [comment]
@Davejkane  @NickHackman 
Layout independent `:` could be resolved with a mixture of keycodes and scancodes.

Use keycodes for detecting when the OS generates a `LSHIFTMOD` or `RSHIFTMOD`. Match the scancode for a `;` on a QWERTY keyboard. 

[KeyDown event struct](https://rust-sdl2.github.io/rust-sdl2/sdl2/event/enum.Event.html#variant.KeyDown)

From [sdl2 docs](http://wiki.libsdl.org/CategoryKeyboard)
>    SDL_Scancode values are used to represent the physical location of a
> keyboard key on the keyboard.
>
>    SDL_Keycode values are mapped to the current layout of the keyboard and
> correlate to an SDL_Scancode.

>  Which one to use is left to the application: scancodes are suited in
> situations where controls are layout-dependent (eg. the "WASD" keys as
> left-handed arrow keys), whereas keycodes are better suited to situations where
> controls are character-dependent (eg. the "I" key for Inventory).

**Comment by NickHackman** [comment]
@Davejkane Little confused what you're referring to, what operations were being unprocessed?

@gurgalex Thanks

Thanks so much for the input it has helped monumentally, and sorry that this has taken so long :heart:

**Comment by Davejkane** [comment]
I'm not sure I'm with you there @gurgalex. The problem isn't that semi-colon is sometimes in a different place, the problem is that colon is not always shift+semi-colon. For example, on a French keyboard (AZERTY) it is an unshifted key, whereas on a German keyboard (QWERTZ), it is shift+comma. So I'm not sure what you are suggesting would work.

**Comment by NickHackman** [comment]
Yeah, going to use [scancode](https://docs.rs/sdl2/0.32.2/sdl2/keyboard/enum.Keycode.html#method.from_scancode) which uses the virtual mapping, going to convert that then check to see if it matches afterwards. I think that's keyboard independent.

**Comment by Davejkane** [comment]
@NickHackman I see that I've misunderstood something there with the unprocessed actions. So just ignore that comment. Just the colon key to handle and then I'm happy.

**Comment by gurgalex** [comment]
@Davejkane 
>I'm not sure I'm with you there @gurgalex. The problem isn't that semi-colon is sometimes in a different place, the problem is that colon is not always shift+semi-colon. For example, on a French keyboard (AZERTY) it is an unshifted key, whereas on a German keyboard (QWERTZ), it is shift+comma. So I'm not sure what you are suggesting would work.

Right, so scancodes return the QWERTY key which corresponds to the physical location of the key pressed regardless of OS keyboard mapping. So, a keypress on an AZERTY or QWERTZ in the same physical location will return the same scancode. (Assuming the hardware mapping isn't different)

**Comment by Davejkane** [comment]
Actually, one last thing! If the bar is hidden because it is toggled off with the `t` key, then entering command mode should force it to be rendered.

**Comment by Davejkane** [comment]
Also, something has been broken with move and delete image. The length of the images vector isn't updated, so the program crashes when moving to the last image. I'm going to have to do a bit more QA on this to see if anything else is broken.

**Comment by gurgalex** [comment]
@NickHackman 

the following results in a panic.

Enter command mode (:)

type `ng` (no space)

Press Enter

```
thread 'main' panicked at 'byte index 3 is out of bounds of `ng`', src/libcore/str/mod.rs:2010:9
```

**Comment by Davejkane** [comment]
@gurgalex, I'm still not sure I understand what you mean. Are you suggesting that a German user types in `Ö` to enter command mode? Because that would be what they would be doing to use the key that is physically in place where an American keyboard has semi-colon.

**Comment by gurgalex** [comment]
>  Are you suggesting that a German user types in Ö to enter command mode? Because that would be what they would be doing to use the key that is physically in place where an American keyboard has semi-colon.

@Davejkane 
Yes.

**Comment by Davejkane** [comment]
Also, @NickHackman no stress and no rush at all. It's totally fine and normal that things break on a big change like this.

**Comment by Davejkane** [comment]
@gurgalex I think a German should type colon like everyone else to get into command mode, otherwise we'll be making lots of user manuals.

**Review by gurgalex — CHANGES_REQUESTED** [review]
Fix crash outlined in https://github.com/Davejkane/riv/pull/58#issuecomment-494526131

**Comment by NickHackman** [comment]
@gurgalex That's not reproducible on my end. Maybe pull again, that was an issue created by me messing up the merging, I fixed that in the latest push.

**Comment by gurgalex** [comment]
> @gurgalex I think a German should type colon like everyone else to get into command mode, otherwise we'll be making lots of user manuals.

@NickHackman 
Okay, in that case. Matching on `TextInput` {text: ":", ..} should work instead.

https://docs.rs/sdl2/0.32.2/sdl2/event/enum.Event.html#variant.TextInput

**Review by gurgalex — APPROVED** [review]

**Comment by gurgalex** [comment]
@NickHackman
You're right. I wasn't using the latest changes.

> @gurgalex That's not reproducible on my end. Maybe pull again, that was an issue created by me messing up the merging, I fixed that in the latest push.

File: src/ui.rs
Comment (NickHackman) [inline src/ui.rs]: Added virtual mapping handled here, but this doesn't actually work on my end, the keycode produced is "semicolon"

So now there is handling for semicolon, colon, and virtual mapping
as mentioned [here](https://wiki.libsdl.org/SDL_Keycode) this just says that the colon isn't physically mapped

**Review by NickHackman — COMMENTED** [review]
This feels like a bad approach, but it covers all the possible bases. It would have weird side effects, as mentioned above by @Davejkane `Ö` would also enter you into command mode, which is undesirable... 

Is the only approach that works is changing everything to be interpreted as `text_input`?

File: src/ui.rs
Comment (Davejkane) [reply src/ui.rs]: What about plain old colon as keycode https://rust-sdl2.github.io/rust-sdl2/sdl2/keyboard/enum.Keycode.html#variant.Colon . Def doesn't work?

Comment (NickHackman) [reply src/ui.rs]: Yeah, that doesn't work on my end nor does `KpColon` wish Colon worked though

Comment (Davejkane) [reply src/ui.rs]: text input does seem like an ugly hack, but it might be the best bet. FYI, I was too lazy to handle this earlier, which is why I went with H for help instead of question mark like I originally planned. So thanks for taking this on.

Comment (NickHackman) [reply src/ui.rs]: Okay, I'm on it

Comment (Davejkane) [reply src/ui.rs]: Hey, and seeing as you have an open pull request, can you please put the stuff about sorting below the stuff about basic controls in the README :)

**Comment by Davejkane** [comment]
So sorry @NickHackman, but the command for changing the destination folder isn't handling tilde properly. It creates a folder in the current directory called `~` which is difficult to `cd` into 😂

**Comment by NickHackman** [comment]
@Davejkane oh dang, my bad, fixing that asap

File: src/ui.rs
Comment (gurgalex) [inline src/ui.rs]: We can probably get rid of tracking `left_shift` and `right_shift` now that textinput will give the character ex. "G"

File: src/ui.rs
Comment (gurgalex) [inline src/ui.rs]: Removing `LShift` and `RShift` here as well.

**Comment by gurgalex** [comment]
@NickHackman 

Command mode does not trigger if the infobar is set to hidden.

We may need to have two bars (an infobar and a command bar).

The command bar renders over the infobar (or just the image if hidden).

**Comment by gurgalex** [comment]
Nevermind, it does trigger when hidden. Just was confused about what I had inputted (errors were not shown in input due to hidden bar)

Suggestion about a command bar still may be useful for showing input and errors.

**Comment by NickHackman** [comment]
@gurgalex now it for sure is visible, just in case.

**Comment by gurgalex** [comment]
@Davejkane @NickHackman 
Do you think equivalent command mode operations should be repeatable with `.` (such as `:help`) -> Action::Help

**Comment by gurgalex** [comment]
Vim seems to use `@:` for repeating the last operation in command mode.
https://vim.fandom.com/wiki/Repeat_last_change

**Comment by gurgalex** [comment]
We could address this later though.

**Comment by NickHackman** [comment]
@gurgalex that sounds like maybe a different PR, but I like the idea

**Comment by Davejkane** [comment]
@NickHackman wow, I'm impressed you got through so many tweaks. One of the big ones left is still that when I move or delete an image, the len of the images vector isn't being updated, meaning that then visiting the last file causes an out of bounds panic.

**Comment by NickHackman** [comment]
@Davejkane I don't think I wrote that bug, but fixed :+1:

edit: Yeah, sorry that is certainly my fault

**Comment by Davejkane** [comment]
It was working fine before the new `max_viewable` logic. And as far as I can tell, it isn't fixed. I just tried to delete the last image when I had 4 images, and it crashed the program.

**Comment by NickHackman** [comment]
@Davejkane I'm confused, that's not reproducible on my end. 
You updated your local repo after the push right?

**Comment by Davejkane** [comment]
OK, just tried again and that seems to be fine, think pulling the branch didn't work. The program now seems to be exiting with this error message now `Error: "SDL2 error: Text has zero width"` when I hit `q` or `esc`. I promise that's my last gripe.

File: src/program/render.rs
Comment (gurgalex) [inline src/program/render.rs:108]: @NickHackman 

Here is where the code errors out.

This debug code shows that the `information` and `mode` are empty strings when the program starts to exit.

```rust
    fn render_infobar(&mut self) -> Result<(), String> {
        let text = infobar::Text::update(&self.ui_state.mode, &self.paths);
        dbg!((&text.mode, &text.information));
        // Load the filename texture
        let filename_surface = self
            .screen
            .font
            .render(&text.information)
            .blended(text_color())
            .map_err(|e| e.to_string()).expect("80: error");

```

File: src/infobar.rs
Comment (gurgalex) [inline src/infobar.rs]: @NickHackman 
(50) This is the line.

The program renders the infobar for the `Exit` mode which returns an empty string for both `mode` and `information`

**Comment by NickHackman** [comment]
Honestly guys thank you so much for all the help, I didn't handle so many edge cases that would've come back to haunt me in the near future. I think we ironed at least a large majority of them, can never say they're all gone, but all the ones that I've tested for are gone. I'm honestly really happy with the result and the level of code quality :smile: 

It was a joy working with you both on this, and I hope to contribute to the project more in the future :heart:

Probably going to do #53 after this

File: src/program/command_mode.rs
Comment (gurgalex) [inline src/program/command_mode.rs]: @NickHackman 

Sorry to request more changes.

The destination folder should be kept the same unless the user changes it with `df`

I believe the field `current_dir` should be called `base_dir` to say that this is the base directory from which the glob finds images.

**Review by gurgalex — CHANGES_REQUESTED** [review]
@NickHackman 
Thanks for all the help.

Should be the last change request from me.

* Rename paths.`current_dir` to `base_dir`
* Keep selected destination folder the same even if a new glob no longer contains the destination folder.

Example

Selected df
`/home/alex/pictures`

Current glob
`/home/alex/*.png`

New glob
`/home/alex/Downloads/*.png`

Destination folder is still
`/home/alex/pictures`

File: src/program/command_mode.rs
Comment (NickHackman) [reply src/program/command_mode.rs]: In your above example that's how it should function currently. 

The only instance it would change in newglob is if the user hasn't modified the default destination folder. If that's the case we should maintain the invariant, stated in help/README, that the default folder should be `./keep`

I think that would make sense, for instance running riv in `/riv/src/program` initially and going to `~/Pictures` and hitting `move` it'd be really weird if we moved the image to `/riv/src/program/keep` rather than `~/Pictures/keep` I personally would think that's unexpected.

I think that we should probably keep track if the user changes the destination folder in case of the instance where the user actually wants the folder name "keep" in a directory. That seems like a bug/unintended behavior.

**Comment by Davejkane** [comment]
I like the behaviour of the keep folder as it is right now. As a user I'll probably make a keep folder on my desktop and jump around in different folders using new glob, and it's nice that the keep folder doesn't move with me. 

I've just got a last little bit of QA to do before my blessing.

**Comment by NickHackman** [comment]
@Davejkane okay, I'll remove that

**Comment by Davejkane** [comment]
In the readme, it says that the command `? or help` will toggle fullscreen. Please correct that. 

~I'm also a little surprised that toggling the help bar with h, and toggling it with the command mode aren't linked, so one can't undo the other. Also, the plain `h` command no longer works when there are no images loaded on initial load, even if I then newglob to get a bunch of pictures, h doesn't work.~

Scrap that, hadn't seen you fixed it to question mark. 👍

**Review by Davejkane — APPROVED** [review]
You did good. Well done. That was a bit epic, but we've pulled through. If there are any bugs, we'll catch them in follow up tickets, but it got past my QA for now. Now that we have this functionality, I'm going to write a couple of tickets.

File: README.md
Comment (gurgalex) [inline README.md:79]: Toggle help menu?

Comment (Davejkane) [reply README.md:79]: fuck that, I can fix it directly. Let's just get this merged.

**Comment by Davejkane** [comment]
@gurgalex are you happy with this PR now? (please approve if so)

**Comment by gurgalex** [comment]
@Davejkane 

Doing some testing to see if this is functioning.

>I like the behaviour of the keep folder as it is right now. As a user I'll probably make a keep folder on my desktop and jump around in different folders using new glob, and it's nice that the keep folder doesn't move with me.

**Comment by gurgalex** [comment]
Seems fine to me. Even though I'm not seeing a commit that undid the changes for a moving `./keep` directory.

**Review by gurgalex — APPROVED** [review]

**Comment by Davejkane** [comment]
I'll do the merge.

**Pull request merged**
